cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/submit/MasterEnvironmentSparkSubmitter.java

```diff
@@ -99,16 +101,36 @@ protected Iterable<LocalizeResource> getArchives(List<LocalizeResource> localize
   }
 
   @Override
-  protected Map<String, String> generateSubmitConf() throws Exception {
-    Map<String, String> config = new HashMap<>();
+  protected Map<String, String> generateSubmitConf(Map<String, String> appConf) throws Exception {
+    Map<String, String> config = new HashMap<>(appConf);
     config.put(SparkConfig.DRIVER_ENV_PREFIX + "CDAP_LOG_DIR", ApplicationConstants.LOG_DIR_EXPANSION_VAR);
     config.put("spark.executorEnv.CDAP_LOG_DIR", ApplicationConstants.LOG_DIR_EXPANSION_VAR);
-    config.put("spark.executorEnv.ARTIFACT_FECTHER_PORT",
-               cConf.get(io.cdap.cdap.app.runtime.spark.Constant.Spark.ArtifactFetcher.PORT));
-    config.putAll(generateOrGetSparkConfig().getConfigs());
+    config.put("spark.executorEnv.ARTIFACT_FECTHER_PORT", cConf.get(ArtifactFetcher.PORT));
+
+    SparkConfig environmentConfig = generateOrGetSparkConfig();
+    config.putAll(mergeSparkConfs(appConf, environmentConfig));
+    return config;
+  }
+
+  @VisibleForTesting
+  static Map<String, String> mergeSparkConfs(Map<String, String> appConf, SparkConfig envConf) {
```

Comment:
clarification: Any particular reason this logic for generating spark submit configurations isn't uniform for different spark submitter implementations?

Reply:
it's more of a historical thing rather than an intended design decision. I think ideally we would have just this MasterEnvironmentSparkSubmitter, but the old yarn stuff never got moved to the MasterEnvironment interface.

Reply:
Thanks!